### PR TITLE
Filtering for the MC truth of V0, photons and cascades for the Nano

### PR DESCRIPTION
**Pull request by a-mathis**
o implementation of a AliAnalysisCut class for the MC stack
o matchig of the photons and v0
o to be done: cascades

The filtering can be steered from the AddTask e.g.
AliAnalysisNanoAODMCParticleCuts* mcCuts = new AliAnalysisNanoAODMCParticleCuts;
mcCuts->SetMinPt(0.1);
mcCuts->SetMaxEta(2);
mcCuts->SetKeepParticles( {333, 3212} );
mcCuts->SetKeepV0s( {3122} );
filterTask->FilterMCStack(mcCuts);

### CODE REVIEW

**Comment by a-mathis** [comment]
@jgrosseo Maybe already now have a look, the cascade cuts are still missing (@bhohlweger will take care of this)

File: PWG/DevNanoAOD/AliAnalysisNanoAODCuts.cxx [hunk 1]
```diff
@@ -734,6 +736,44 @@ Bool_t AliAnalysisNanoAODEventCuts::IsSelected(TObject* obj)
...
+      fMinPt(0.f),
+      fMaxEta(999.f),
+      fPDGToKeep(),
+      fPDGV0() {
+}
+
+bool AliAnalysisNanoAODMCParticleCuts::IsSelected(TObject* obj) {
+  // Returns true if the MC particle is to be kept!
+  AliAODMCParticle* part = dynamic_cast<AliAODMCParticle*>(obj);
+
```
Comment (jgrosseo) [inline PWG/DevNanoAOD/AliAnalysisNanoAODCuts.cxx]: Add AliFatal if part == 0 after the cast

File: PWG/DevNanoAOD/AliAnalysisNanoAODCuts.cxx [hunk 2]
```diff
@@ -734,6 +736,44 @@ Bool_t AliAnalysisNanoAODEventCuts::IsSelected(TObject* obj)
...
+    return false;
+  }
+  if (TMath::Abs(part->Eta()) > fMaxEta) {
+    return false;
+  }
+  if (fDoSelectPrimaries && !part->IsPhysicalPrimary()) {
+    return false;
+  }
+  if (fDoSelectCharged && !part->Charge()) {
+    return false;
```
Comment (jgrosseo) [inline PWG/DevNanoAOD/AliAnalysisNanoAODCuts.cxx]: Is the ! clean for comparing to -1 ?

File: PWG/DevNanoAOD/AliAnalysisNanoAODCuts.cxx [hunk 3]
```diff
@@ -734,6 +736,44 @@ Bool_t AliAnalysisNanoAODEventCuts::IsSelected(TObject* obj)
...
+      fPDGToKeep(),
+      fPDGV0() {
+}
+
+bool AliAnalysisNanoAODMCParticleCuts::IsSelected(TObject* obj) {
+  // Returns true if the MC particle is to be kept!
+  AliAODMCParticle* part = dynamic_cast<AliAODMCParticle*>(obj);
+
+  // if this is activated the specified particles are kept ignoring the cuts
+  if (fForceKeepParticles) {
```
Comment (jgrosseo) [inline PWG/DevNanoAOD/AliAnalysisNanoAODCuts.cxx]: Why is this needed? For performance? If the list is empty (default), it will just continue...

File: PWG/DevNanoAOD/AliAnalysisNanoAODCuts.h [hunk 4]
```diff
@@ -223,6 +223,65 @@ class AliAnalysisNanoAODEventCuts : public AliAnalysisCuts
...
+    fPDGToKeep = pdgCodes;
+    fForceKeepParticles = true;
+  }
+  std::vector<int> &GetKeepV0s() {
+    return fPDGV0;
+  }
+  void SetKeepV0s(std::vector<int> pdgCodes) {
+    fPDGV0 = pdgCodes;
+  }
+
```
Comment (jgrosseo) [inline PWG/DevNanoAOD/AliAnalysisNanoAODCuts.h]: I prefer to make all the setters/getters in the header file one liners, to make it more compact. What do you think?

File: PWG/DevNanoAOD/AliAnalysisNanoAODCuts.h [hunk 5]
```diff
@@ -223,6 +223,65 @@ class AliAnalysisNanoAODEventCuts : public AliAnalysisCuts
...
+  void SetMaxEta(float var) {
+    fMaxEta = var;
+  }
+  std::vector<int> &GetKeepParticles() {
+    return fPDGToKeep;
+  }
+  void SetKeepParticles(std::vector<int> pdgCodes) {
+    fPDGToKeep = pdgCodes;
+    fForceKeepParticles = true;
+  }
```
Comment (jgrosseo) [inline PWG/DevNanoAOD/AliAnalysisNanoAODCuts.h]: In addition, let's add a function AddKeepParticles(int pdgCode) which adds it to the list

File: PWG/DevNanoAOD/AliNanoAODReplicator.h [hunk 6]
```diff
@@ -77,6 +78,7 @@ class AliNanoAODReplicator : public AliAODBranchReplicator
   void SetV0Cuts(AliAnalysisCuts* cuts) { fV0Cuts = cuts; }
   void SetCascadeCuts(AliAnalysisCuts* cuts) { fCascadeCuts = cuts; }
   void SetConversionPhotonCuts(AliAnalysisCuts* cuts) { fConversionPhotonCuts = cuts; }
+  void SetMCParticleCuts(AliAnalysisNanoAODMCParticleCuts* cuts) { fMCParticleCuts = cuts; }
```
Comment (jgrosseo) [inline PWG/DevNanoAOD/AliNanoAODReplicator.h]: Please keep this also AliAnalysisCuts*?

File: PWG/DevNanoAOD/AliNanoAODReplicator.h [hunk 7]
```diff
@@ -99,13 +101,17 @@ class AliNanoAODReplicator : public AliAODBranchReplicator
...
   void FilterMC(const AliAODEvent& source);
   AliAODVertex* CloneAndStoreVertex(AliAODVertex* toClone);
  
   AliAnalysisCuts* fTrackCuts; // decides which tracks to keep
   AliAnalysisCuts* fV0Cuts;    // decides which V0s to keep
   AliAnalysisCuts* fCascadeCuts; // decides which cascades to keep
   AliAnalysisCuts* fConversionPhotonCuts; // decides which conversion photons to keep
+  AliAnalysisNanoAODMCParticleCuts* fMCParticleCuts;  // decides which particles from the MC stack to keep
+                                                      // Here we need the actual object because we retrieve the MC
+                                                      // matching of the V0s from here
```
Comment (jgrosseo) [inline PWG/DevNanoAOD/AliNanoAODReplicator.h]: Same here

Comment (a-mathis) [reply PWG/DevNanoAOD/AliNanoAODReplicator.h]: I know, but please read the comment ;)
Same for your comment above

File: PWG/DevNanoAOD/AliAnalysisNanoAODCuts.cxx (hunk 2 above)
Comment (a-mathis) [reply PWG/DevNanoAOD/AliAnalysisNanoAODCuts.cxx]: yes, it just checks == 0 of != 0
so this should be fine

File: PWG/DevNanoAOD/AliNanoAODReplicator.h (hunk 7 above)
Comment (a-mathis) [reply PWG/DevNanoAOD/AliNanoAODReplicator.h]: ok, wenn ich das in der cxx caste funktioniert das auch mit nem AliAnalysisCuts object! cool

**Comment by a-mathis** [comment]
Hi @jgrosseo,
I modified according to your comments and did some measurements how much the MC information contributes to the total size of the NanoAOD

- Mode 1: 91 %
- Mode 2 (only charged primaries and the tracks, v0s, ...): 54 %
- Mode 2 (additionally requiring pT > 0.1, |eta| < 2 on the primaries): 39 %
- Mode 2 (additionally requiring pT > 0.15, |eta| < 1.5 on the primaries): 36 %
- Mode 2 (additionally keeping all phi and Sigma0): 37 %

From my point of view this could be merged now, so I will remove the WIP flag

**Comment by jgrosseo** [comment]
+1
